### PR TITLE
Channel AddSequencedLeaves to storage

### PR DESCRIPTION
**Pull request by pavelkalinnikov**
This change adds `AddSequencedLeaves` to the storage API and implements the corresponding public API on top of it. This also adds method stubs to concrete storage implementations.

### CODE REVIEW

File: server/log_rpc_server.go [hunk 1]
```diff
@@ -130,13 +137,56 @@ func (t *TrillianLogRPCServer) QueueLeaves(ctx context.Context, req *trillian.Qu
...
+		Leaves: []*trillian.LogLeaf{req.Leaf},
+	}
+	rsp, err := t.AddSequencedLeaves(ctx, batchReq)
+	if err != nil {
+		return nil, err
+	}
+	if rsp == nil {
+		return nil, status.Errorf(codes.Internal, "missing response")
+	}
+	if ln := len(rsp.Results); ln != 1 {
```
Comment (Martin2112) [inline server/log_rpc_server.go]: Maybe do got, want here.

File: server/log_rpc_server.go [hunk 2]
```diff
@@ -130,13 +137,56 @@ func (t *TrillianLogRPCServer) QueueLeaves(ctx context.Context, req *trillian.Qu
...
+	if err := validateAddSequencedLeavesRequest(req); err != nil {
+		return nil, err
+	}
+
+	tree, hasher, err := t.getTreeAndHasher(ctx, req.LogId, false /* readonly */)
+	if err != nil {
+		return nil, err
+	}
+	if want := trillian.TreeType_PREORDERED_LOG; tree.TreeType != want {
+		return nil, status.Errorf(codes.FailedPrecondition, "adding sequenced leaves available only for %v trees", want)
```
Comment (Martin2112) [inline server/log_rpc_server.go]: Error is a bit wordy and consider adding what state the tree is in to help the caller.

File: server/validate.go [hunk 3]
```diff
@@ -106,28 +106,33 @@ func validateGetEntryAndProofRequest(req *trillian.GetEntryAndProofRequest) erro
...
 	}
 	return nil
 }
 
 func validateQueueLeavesRequest(req *trillian.QueueLeavesRequest) error {
-	if len(req.Leaves) == 0 {
-		return status.Error(codes.InvalidArgument, "QueueLeavesRequest.Leaves empty")
+	if err := validateLogLeaves(req.Leaves); err != nil {
+		// validateLogLeaves errors chain nicely with "QueueLeavesRequest.".
+		return status.Errorf(codes.InvalidArgument, "QueueLeavesRequest.", err)
```
Comment (Martin2112) [inline server/validate.go]: Errorf() with nothing in the format string. Is something missing here?

File: storage/log_storage.go [hunk 4]
```diff
@@ -124,6 +125,30 @@ type LogStorage interface {
 	// Duplicates are only reported if the underlying tree does not permit duplicates, and are
 	// considered duplicate if their leaf.LeafIdentityHash matches.
 	QueueLeaves(ctx context.Context, treeID int64, leaves []*trillian.LogLeaf, queueTimestamp time.Time) ([]*trillian.QueuedLogLeaf, error)
+
+	// AddSequencedLeaves stores the `leaves` and associates them with the log
+	// positions according to their `LeafIndex` field. The indices must be unique
+	// and sorted in ascending order.
```
Comment (Martin2112) [inline storage/log_storage.go]: Where is the ordering constraint enforced?

File: server/log_rpc_server.go [hunk 5]
```diff
@@ -130,13 +137,56 @@ func (t *TrillianLogRPCServer) QueueLeaves(ctx context.Context, req *trillian.Qu
...
 
 // AddSequencedLeaves submits a batch of sequenced leaves to a pre-ordered log
 // for later integration into its underlying tree.
 func (t *TrillianLogRPCServer) AddSequencedLeaves(ctx context.Context, req *trillian.AddSequencedLeavesRequest) (*trillian.AddSequencedLeavesResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "AddSequencedLeaves is not implemented")
+	if err := validateAddSequencedLeavesRequest(req); err != nil {
+		return nil, err
+	}
+
+	tree, hasher, err := t.getTreeAndHasher(ctx, req.LogId, false /* readonly */)
```
Comment (pavelkalinnikov) [inline server/log_rpc_server.go]: This line will not get the tree currently, as `getTreeAndHasher` uses `GetOpts` with `LOG` type. Need to refactor it a bit.

File: server/validate.go [hunk 6]
```diff
@@ -106,28 +106,33 @@ func validateGetEntryAndProofRequest(req *trillian.GetEntryAndProofRequest) erro
 }
 
 func validateQueueLeafRequest(req *trillian.QueueLeafRequest) error {
-	if req.Leaf == nil {
-		return status.Error(codes.InvalidArgument, "QueueLeafRequest.Leaf empty")
+	if err := validateLogLeaf(req.Leaf); err != nil {
```
Comment (Martin2112) [inline server/validate.go]: Could we have a validateLeaf(leaf, "QueueLeaf") function and get rid of some of the duplication? They're all a bit similar.

File: server/log_rpc_server.go (hunk 5 above)
Comment (pavelkalinnikov) [reply server/log_rpc_server.go]: Time to add some tests...

File: server/validate.go [hunk 7]
```diff
@@ -106,28 +106,33 @@ func validateGetEntryAndProofRequest(req *trillian.GetEntryAndProofRequest) erro
...
-			// validateLogLeaf errors are meant to chain nicely with "Leaves."
-			return status.Errorf(codes.InvalidArgument, "QueueLeavesRequest.Leaves[%v].%v", i, err)
-		}
+	return nil
+}
+
+func validateAddSequencedLeavesRequest(req *trillian.AddSequencedLeavesRequest) error {
+	if err := validateLogLeaves(req.Leaves); err != nil {
+		// validateLogLeaves errors chain nicely with "AddSequencedLeavesRequest.".
+		return status.Errorf(codes.InvalidArgument, "AddSequencedLeavesRequest.", err)
```
Comment (Martin2112) [inline server/validate.go]: Similar here, Errorf() with no formatting.

File: server/validate.go [hunk 8]
```diff
@@ -106,28 +106,33 @@ func validateGetEntryAndProofRequest(req *trillian.GetEntryAndProofRequest) erro
...
+		return status.Errorf(codes.InvalidArgument, "AddSequencedLeafRequest.Leaf%v", err)
 	}
 	return nil
 }
 
 func validateQueueLeavesRequest(req *trillian.QueueLeavesRequest) error {
-	if len(req.Leaves) == 0 {
-		return status.Error(codes.InvalidArgument, "QueueLeavesRequest.Leaves empty")
+	if err := validateLogLeaves(req.Leaves); err != nil {
+		// validateLogLeaves errors chain nicely with "QueueLeavesRequest.".
```
Comment (Martin2112) [inline server/validate.go]: Can I just check that what I understand by "chain" is what you mean?

File: server/validate.go (hunk 6 above)
Comment (pavelkalinnikov) [reply server/validate.go]: Done. Also did the same in `validateLogLeaves`.

**Comment by pavelkalinnikov** [comment]
`server/log_rpc_server_test.go` file is marked as generated for some reason. Please expand and review it as well.

File: server/validate.go [hunk 9]
```diff
@@ -106,47 +106,61 @@ func validateGetEntryAndProofRequest(req *trillian.GetEntryAndProofRequest) erro
...
+
+	lastIndex := int64(-1)
 	for i, leaf := range req.Leaves {
-		if leaf == nil {
-			return status.Errorf(codes.InvalidArgument, "QueueLeavesRequest.Leaves[%v] empty", i)
-		}
-		if err := validateLogLeaf(leaf); err != nil {
-			// validateLogLeaf errors are meant to chain nicely with "Leaves."
-			return status.Errorf(codes.InvalidArgument, "QueueLeavesRequest.Leaves[%v].%v", i, err)
+		if leaf.LeafIndex <= lastIndex {
```
Comment (AlCutter) [inline server/validate.go]: Do we want to ensure the leaves are contiguous?

File: storage/log_storage.go [hunk 10]
```diff
@@ -124,6 +125,30 @@ type LogStorage interface {
...
+	//
+	// If error is nil, the returned slice is the same size as the input, entries
+	// correspond to the `leaves` in the same order. Each entry describes the
+	// result of adding the corresponding leaf.
+	//
+	// Possible `QueuedLogLeaf.status` values with their semantics:
+	//  - OK: The leaf has been successfully stored.
+	//  - AlreadyExists: The storage has another leaf with the same `LeafIndex`.
```
Comment (AlCutter) [inline storage/log_storage.go]: "the storage already contains an identical leaf at the specified `LeafIndex`."?

File: server/validate.go (hunk 9 above)
Comment (pavelkalinnikov) [reply server/validate.go]: Initially I wanted do so, but I was not sure.

Suppose you submit entries 0-99, but get a response that only entries 0,2,4,...,98 were saved. You can further submit entries 1,3,...,99 in a single batch if non-contiguous is allowed. If not, you can do one of:

extreme1: 50 one-element ranges (kills the advantage of batching)
extreme2: all 100 entries (50 entries are re-submitted)
trade-off between the two (e.g., [0-3] [5-8] ...)
I am not sure how much the above example is realistic.

Perhaps we can start with an API with tighter requirements (that is, submit only contiguous ranges). If it ends up too strict, the requirements can be loosened. WDYT?

File: storage/log_storage.go (hunk 10 above)
Comment (pavelkalinnikov) [reply storage/log_storage.go]: Done.

File: server/validate.go (hunk 9 above)
Comment (pavelkalinnikov) [reply server/validate.go]: @gdbelvin In Key Transparency, do you ever need to submit arbitrary non-contiguous batches? Or single entry / contiguous batches would be sufficient?

Comment (pavelkalinnikov) [reply server/validate.go]: @AlCutter I propose to fine-tune this API in follow-up PRs, as I get more insight into what we can do with the batches in MySQL.

**Comment by pavelkalinnikov** [comment]
@Martin2112 Can you take another look today?

**Comment by Martin2112** [comment]
I'll try to make time to look at it again.

File: server/log_rpc_server.go [hunk 11]
```diff
@@ -504,7 +557,7 @@ func (t *TrillianLogRPCServer) getTreeAndHasher(ctx context.Context, treeID int6
 // InitLog initialises a freshly created Log by creating the first STH with size 0.
 func (t *TrillianLogRPCServer) InitLog(ctx context.Context, req *trillian.InitLogRequest) (*trillian.InitLogResponse, error) {
 	logID := req.LogId
-	tree, hasher, err := t.getTreeAndHasher(ctx, logID, false /* readonly */)
+	tree, hasher, err := t.getTreeAndHasher(ctx, logID, optsLogRead)
```
Comment (Martin2112) [inline server/log_rpc_server.go]: write?

File: server/validate.go [hunk 12]
```diff
@@ -106,47 +106,62 @@ func validateGetEntryAndProofRequest(req *trillian.GetEntryAndProofRequest) erro
...
+}
+
+func validateAddSequencedLeafRequest(req *trillian.AddSequencedLeafRequest) error {
+	return validateLogLeaf(req.Leaf, "AddSequencedLeafRequest.Leaf")
 }
 
 func validateQueueLeavesRequest(req *trillian.QueueLeavesRequest) error {
-	if len(req.Leaves) == 0 {
-		return status.Error(codes.InvalidArgument, "QueueLeavesRequest.Leaves empty")
+	return validateLogLeaves(req.Leaves, "QueueLeavesRequest")
```
Comment (Martin2112) [inline server/validate.go]: Can we not just remove this and the 2 functions above and call the validateLogXXX() functions directly?

File: server/validate.go [hunk 13]
```diff
@@ -106,47 +106,62 @@ func validateGetEntryAndProofRequest(req *trillian.GetEntryAndProofRequest) erro
...
+	nextIndex := req.Leaves[0].LeafIndex
 	for i, leaf := range req.Leaves {
-		if leaf == nil {
-			return status.Errorf(codes.InvalidArgument, "QueueLeavesRequest.Leaves[%v] empty", i)
-		}
-		if err := validateLogLeaf(leaf); err != nil {
-			// validateLogLeaf errors are meant to chain nicely with "Leaves."
-			return status.Errorf(codes.InvalidArgument, "QueueLeavesRequest.Leaves[%v].%v", i, err)
+		if leaf.LeafIndex != nextIndex {
+			return status.Errorf(codes.OutOfRange, "%v.Leaves[%v].LeafIndex=%v, want %v", prefix, i, leaf.LeafIndex, nextIndex)
```
Comment (Martin2112) [inline server/validate.go]: This is checking for sequential index? Are we using OutOfRange to be able to distinguish it because it could be FailedPrecondition?

File: server/log_rpc_server.go (hunk 11 above)
Comment (pavelkalinnikov) [reply server/log_rpc_server.go]: Yes, fixed.

Actually, `InitLog` should accept `PREORDERED_LOG` in this line as well, but this requires `LogStorage.ReadWriteTransaction` refactoring similar to #1007. Left a TODO.

File: server/validate.go (hunk 12 above)
Comment (pavelkalinnikov) [reply server/validate.go]: Done.

File: server/validate.go (hunk 13 above)
Comment (pavelkalinnikov) [reply server/validate.go]: Right, let's make it a `FailedPrecondition`.

**Comment by pavelkalinnikov** [comment]
@AlCutter ping

File: server/log_rpc_server.go [hunk 14]
```diff
@@ -130,13 +143,53 @@ func (t *TrillianLogRPCServer) QueueLeaves(ctx context.Context, req *trillian.Qu
...
+		return nil, err
+	}
+
+	ctx = trees.NewContext(ctx, tree)
+	leaves, err := t.registry.LogStorage.AddSequencedLeaves(ctx, tree.TreeId, req.Leaves)
+	if err != nil {
+		return nil, err
+	}
+	if got, want := len(leaves), len(req.Leaves); got != want {
+		return nil, status.Errorf(codes.Internal, "AddSequencedLeaves returned %d codes, want: %d", got, want)
```
Comment (AlCutter) [inline server/log_rpc_server.go]: codes->leaves?

Comment (pavelkalinnikov) [reply server/log_rpc_server.go]: Done.
